wandb/integration/yolov8/utils.py

```diff
@@ -0,0 +1,173 @@
+        if len(cls) > 0:
+            img = images[i].transpose(1, 2, 0)
+            image_kwargs = {
+                "data_or_path": img,
+                "caption": pathlib.Path(captions[i]).name,
+            }
+            prediction_boxes_data, prediction_masks_data = load_boxes_data(
+                i, batch_idx, cls, bboxes, masks, names, h, w
+            )
+            ground_truth_boxes_data, ground_truth_masks_data = load_boxes_data(
+                i,
+                batch["batch_idx"].cpu().numpy(),
+                batch["cls"].squeeze(-1).cpu().numpy(),
+                batch["bboxes"].cpu().numpy(),
+                batch["masks"].cpu().numpy()
+                if batch.get("masks") is not None
+                else None,
+                names,
+                h,
+                w,
+            )
+            image_kwargs["boxes"] = {
+                "predictions": {
+                    "box_data": prediction_boxes_data,
+                    "class_labels": names,
+                },
+                "ground_truth": {
+                    "box_data": ground_truth_boxes_data,
+                    "class_labels": names,
+                },
+            }
+            if prediction_masks_data is not None:
+                image_kwargs["masks"] = {
+                    "predictions": {
+                        "mask_data": prediction_masks_data,
+                        "class_labels": names,
+                    }
+                }
+            if ground_truth_masks_data is not None:
+                image_kwargs["masks"]["ground_truth"] = {
+                    "mask_data": ground_truth_masks_data,
+                    "class_labels": names,
+                }
+            image_kwargs["classes"] = class_set
+            object_confidences: Dict[str, Any] = {}
+            for prediction_box in prediction_boxes_data:
+                object_confidences[
+                    prediction_box["box_caption"]
+                ] = object_confidences.get(prediction_box["box_caption"], []) + [
+                    prediction_box["scores"]["conf"]
+                ]
+            for key, value in object_confidences.items():
+                object_confidences[key] = sum(value) / len(value)
+            scores_data.append(object_confidences)
+            ground_truth_kwargs = deepcopy(image_kwargs)
+            ground_truth_kwargs["boxes"].pop("predictions")
+            if ground_truth_kwargs.get("masks") is not None:
+                ground_truth_kwargs["masks"].pop("predictions")
+            # TODO: Consider removing ground truth from predictions
+            # image_kwargs["boxes"].pop("ground_truth")
+            # image_kwargs["masks"].pop("ground_truth")
+            out_images.append(
+                (
+                    wandb.Image(**ground_truth_kwargs),
+                    wandb.Image(**image_kwargs),
+                )
+            )
+
+    return out_images, scores_data
+
+
+def load_boxes_data(
```

Comment:
could you add more documentation this function has many assumptions makes it hard to read

Reply:
also not reason to mix the bounding boxes and masks in the same function. 
I think breaking it into 2 functions will make it more readable